email/blocker.go

```diff
@@ -0,0 +1,215 @@
+func NewBlocker(ctx context.Context, blockerAuthHeader, blockerApiUrl string, database *database.AbuseScannerDB, logger *logrus.Logger) *Blocker {
+	return &Blocker{
+		staticBlockerAuthHeader: blockerAuthHeader,
+		staticBlockerApiUrl:     blockerApiUrl,
+		staticContext:           ctx,
+		staticDatabase:          database,
+		staticLogger:            logger.WithField("module", "Blocker"),
+	}
+}
+
+// Start initializes the blocker process.
+func (b *Blocker) Start() error {
+	go b.threadedBlockMessages()
+	return nil
+}
+
+// threadedBlockMessages will periodically fetch email messages that have not
+// been blocked yet and feed them to the blocker API.
+func (b *Blocker) threadedBlockMessages() {
+	// convenience variables
+	logger := b.staticLogger
+	abuseDB := b.staticDatabase
+	first := true
+
+	// start the loop
+	for {
+		// sleep until next iteration, sleeping at the start of the for loop
+		// allows to 'continue' on error.
+		if !first {
+			select {
+			case <-b.staticContext.Done():
+				logger.Debugln("Blocker context done")
+				return
+			case <-time.After(blockFrequency):
+			}
+		}
+		first = false
+
+		logger.Debugln("Triggered")
+
+		// fetch all unblocked emails
+		toBlock, err := abuseDB.FindUnblocked()
+		if err != nil {
+			logger.Errorf("Failed fetching unparsed emails, error %v", err)
+			continue
+		}
+
+		// log unblocked messages count
+		numUnblocked := len(toBlock)
+		if numUnblocked == 0 {
+			logger.Debugf("Found %v unblocked messages\n", numUnblocked)
+		} else {
+			logger.Infof("Found %v unblocked messages\n", numUnblocked)
+		}
+
+		// loop all emails and parse them
+		for _, email := range toBlock {
+			err = func() (err error) {
+				lock := abuseDB.NewLock(email.UID)
+				err = lock.Lock()
+				if err != nil {
+					return errors.AddContext(err, "could not acquire lock")
+				}
+				defer func() {
+					unlockErr := lock.Unlock()
+					if unlockErr != nil {
+						err = errors.Compose(err, errors.AddContext(unlockErr, "could not release lock"))
+						return
+					}
+				}()
+
+				// parse the email
+				result, err := b.blockReport(email.ParseResult)
+				if err != nil {
+					return errors.AddContext(err, "could not block report")
+				}
+
+				// sanity check we have a result for every skylink
+				if len(result) != len(email.ParseResult.Skylinks) {
+					return errors.New("block result not defined for every skylink")
+				}
+
+				// update the email
+				err = abuseDB.UpdateNoLock(email,
+					bson.D{
+						{"$set", bson.D{
+							{"blocked", true},
+							{"blockResult", result},
+						}},
+					},
+				)
+				if err != nil {
+					return errors.AddContext(err, "could not update email")
+				}
+				return nil
+			}()
+			if err != nil {
+				logger.Errorf("Failed to parse email %v, error %v", email.UID, err)
+			}
+		}
+	}
+}
+
+// blockReport will block all skylinks from the given abuse report.
+func (b *Blocker) blockReport(report database.AbuseReport) ([]string, error) {
```

Comment:
REQ: This method can never return an error. Even when it encounters errors, it returns error `nil`.

And then if the number of results doesn't match the number of skylinks, we don't mark the report as done. But we *do* output a result for each skylink, right? Even when the result is "we failed to block this". 

We need either the anonymous func to return an error, so we don't add the result to the output list if it's not a success or to check whether the result is `BLOCKED`. But if we have that situation then we can just return the number of blocked skylinks and log the errors, as usual.

In any case, (unless I'm missing something) this method will always return as many results as there are skylinks in the report and this is not the intended behaviour.

Reply:
I just have it return `[]string` now. You are right the error was an oversight. This method used to be different.
The goal is to get a string message for every skylink in the abuse report. So if there's 2 skylinks found this should return `["BLOCKED", "BLOCKED"]` in case of success, and `["BLOCKED", "reason_for_failure"]` in case of failure. I know that looks a bit weird but I store it that way because it makes it easy to write an aggregation pipeline in the future. Every Skylink should correspond to an index in the results array.

Reply:
Oh, so when we get a failure we won't retry it later or report it as an error? We'll just mark it down and carry on. OK, I've missed that when I reviewed.

F/U: What's the plan for the failures? We should either retry them repeatedly or send some sort of notification, so we know that we need to manually deal with them or at least debug while there are failures, as it might be an indication of a bug.

In any case, not a blocker in this case - feel free to resolve.